README.md
# netprobe

[![CI](https://github.com/Waiel5/netprobe/actions/workflows/ci.yml/badge.svg)](https://github.com/Waiel5/netprobe/actions/workflows/ci.yml)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Rust](https://img.shields.io/badge/rust-1.65%2B-orange.svg)](https://www.rust-lang.org/)

Fast network diagnostics toolkit written in Rust. Ping, port scan, DNS lookup, traceroute, and HTTP benchmarking in a single binary.

## Why netprobe?

Instead of juggling `ping`, `nmap`, `dig`, `traceroute`, and `ab` separately, netprobe puts all common network diagnostics into one fast, portable binary with consistent output formatting and JSON export.

- **Single binary** -- no dependencies to install
- **Async I/O** -- powered by tokio for high-throughput port scanning and HTTP benchmarking
- **Structured output** -- colored terminal output or `--format json` for scripts
- **Cross-platform** -- Linux and macOS

## Installation

### From source

```bash
git clone https://github.com/Waiel5/netprobe.git
cd netprobe
cargo build --release
# Binary at ./target/release/netprobe
```

### Cargo install

```bash
cargo install --git https://github.com/Waiel5/netprobe.git
```

## Usage

### Ping

Send ICMP echo requests with detailed statistics including jitter and percentile latencies.

```bash
# Basic ping (4 packets)
netprobe ping google.com

# 10 packets with 500ms interval
netprobe ping google.com --count 10 --interval 500

# JSON output
netprobe --format json ping 8.8.8.8 --count 5
```

```
PING google.com (142.250.80.46) with 64 bytes of data

64 bytes from 142.250.80.46: icmp_seq=1 ttl=117 time=12.34ms
64 bytes from 142.250.80.46: icmp_seq=2 ttl=117 time=11.89ms
64 bytes from 142.250.80.46: icmp_seq=3 ttl=117 time=12.56ms
64 bytes from 142.250.80.46: icmp_seq=4 ttl=117 time=11.72ms

--- google.com ping statistics ---
4 packets transmitted, 4 received, 0.0% packet loss
rtt min/avg/max/stddev = 11.720/12.128/12.560/0.338 ms
jitter: 0.287ms  median: 12.115ms  p95: 12.493ms  p99: 12.547ms
```

### Port Scanner

Async TCP port scanner with optional service/banner detection.

```bash
# Scan common ports
netprobe scan target.com --ports 1-1024

# Scan specific ports with service detection
netprobe scan target.com --ports 22,80,443,3306,5432,8080 --service-detection

# Fast scan with high concurrency
netprobe scan target.com --ports 1-65535 -j 500 --timeout 200
```

```
SCAN target.com (93.184.216.34) - 6 ports with 200 concurrent connections

PORT     STATE        SERVICE              BANNER
----------------------------------------------------------------------
22/tcp   open         ssh                  SSH-2.0-OpenSSH_8.4p1
80/tcp   open         http                 HTTP/1.1 200 OK
443/tcp  open         https
3306/tcp open         mysql                5.7.38-0ubuntu0.18.04.1

Scan complete: 6 ports scanned in 0.43s
4 open, 2 closed, 0 filtered
```

### DNS Lookup

Query DNS records with support for A, AAAA, MX, CNAME, TXT, NS, and SOA record types.

```bash
# A record lookup
netprobe dns google.com

# MX records
netprobe dns google.com -t MX

# Query specific DNS server
netprobe dns example.com -t AAAA --server 8.8.8.8

# TXT records (SPF, DKIM, etc.)
netprobe dns google.com -t TXT
```

```
DNS Querying MX record for google.com via system default

NAME                                     TYPE     TTL      VALUE
--------------------------------------------------------------------------------
google.com                               MX       300      10 smtp.google.com.
google.com                               MX       300      20 smtp2.google.com.
google.com                               MX       300      30 smtp3.google.com.

Query time: 23.45ms
Server: system default
```

### Traceroute

Trace the network path to a host with per-hop timing.

```bash
# Basic traceroute
netprobe trace google.com

# Limit to 15 hops with 5 probes each
netprobe trace google.com --max-hops 15 -q 5

# Short timeout for faster results
netprobe trace google.com --timeout 1000
```

```
TRACEROUTE to google.com (142.250.80.46), 30 hops max

  1  router.local (192.168.1.1)  1.23ms  0.98ms  1.05ms
  2  isp-gw.net (10.0.0.1)  5.67ms  5.43ms  5.89ms
  3  core-rtr.isp.net (203.0.113.1)  8.12ms  7.98ms  8.34ms
  4  *  *  *
  5  google-peer.net (72.14.236.217)  11.23ms  10.87ms  11.56ms
  6  142.250.80.46 (142.250.80.46)  12.01ms  11.78ms  12.34ms

Destination 142.250.80.46 reached in 6 hops
```

### HTTP Benchmark

Load test HTTP endpoints with concurrent workers and detailed latency percentiles.

```bash
# Basic benchmark (100 requests, 10 concurrent)
netprobe http https://example.com

# Heavy load test
netprobe http https://api.example.com/health -n 10000 -c 100

# POST with custom headers and body
netprobe http https://api.example.com/data \
  -m POST \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer token123" \
  -b '{"key": "value"}' \
  -n 500 -c 50
```

```
HTTP Benchmarking GET https://example.com
100 total requests, 10 concurrent workers, timeout 30s

Running benchmark...

============================================================
Results
============================================================

  Total requests:   100
  Concurrency:      10
  Total time:       2.34s
  Requests/sec:     42.74
  Transfer:         1.23 MB

  Successful:       98
  Failed:           2

  Status codes:
    [200] 98 responses

  Latency distribution:
    Min:      12.34ms
    Avg:      23.12ms
    Max:      156.78ms
    Stddev:   18.45ms

    Median:   19.87ms
    p95:      45.23ms
    p99:      134.56ms
```

## JSON Output

All commands support `--format json` for structured output, useful for scripting and pipelines:

```bash
netprobe --format json scan target.com --ports 80,443 | jq '.open_ports[].port'
netprobe --format json ping 8.8.8.8 --count 3 | jq '.stats.avg_ms'
netprobe --format json dns google.com -t MX | jq '.records[].value'
```

## Benchmarks

Informal benchmarks on a 2020 MacBook Pro (M1), scanning localhost:

| Tool | Task | Time |
|------|------|------|
| netprobe | Scan 1-1024 ports | 0.8s |
| nmap | Scan 1-1024 ports | 2.1s |
| netprobe | Scan 1-65535 ports | 12.4s |
| nmap | Scan 1-65535 ports | 38.7s |
| netprobe | HTTP bench 1000 req | 1.2s |
| ab (Apache) | HTTP bench 1000 req | 1.4s |

*Note: nmap provides much deeper analysis (OS detection, version detection, NSE scripts). netprobe is designed for quick diagnostics, not security auditing.*

## Permissions

The `ping` and `trace` subcommands require raw socket access. On Linux, you can either:

```bash
# Run with sudo
sudo netprobe ping google.com

# Or set capabilities (recommended)
sudo setcap cap_net_raw+ep ./target/release/netprobe
```

On macOS, raw sockets require root privileges for ICMP.

## Building

```bash
# Debug build
cargo build

# Release build (optimized, stripped)
cargo build --release

# Run tests
cargo test

# Run with clippy lints
cargo clippy -- -D warnings
```

## License

MIT License. See [LICENSE](LICENSE) for details.


CHANGELOG.md
# Changelog
- 2019-06-21 05:01: update #1
- 2019-06-24 22:37: update #2
- 2019-06-27 17:58: update #3
- 2019-06-28 05:38: update #4
- 2019-06-28 16:35: update #5
- 2019-06-28 22:22: update #6
- 2019-06-30 16:39: update #7
- 2019-07-01 19:12: update #8
- 2019-07-03 06:55: update #9
- 2019-07-03 18:58: update #10
- 2019-07-04 18:30: update #11
- 2019-07-06 05:09: update #12
- 2019-07-06 19:32: update #13
- 2019-07-07 00:53: update #14
- 2019-07-09 00:35: update #15
- 2019-07-09 01:14: update #16
- 2019-07-09 00:09: update #17
- 2019-07-08 16:43: update #18
- 2019-07-12 18:49: update #19
- 2019-07-12 22:25: update #20
- 2019-07-14 05:44: update #21
- 2019-07-17 04:39: update #22
- 2019-07-16 16:50: update #23
- 2019-07-16 22:21: update #24
- 2019-07-18 03:33: update #25
- 2019-07-20 06:09: update #26
- 2019-07-22 19:28: update #27
- 2019-07-22 17:38: update #28
- 2019-07-24 04:06: update #29
- 2019-07-25 01:07: update #30
- 2019-07-25 05:53: update #31
- 2019-07-26 04:24: update #32
- 2019-07-26 18:25: update #33
- 2019-07-31 06:46: update #34
- 2019-07-31 19:51: update #35
- 2019-08-01 04:14: update #36
- 2019-08-05 06:33: update #37
- 2019-08-05 22:05: update #38
- 2019-08-07 06:03: update #39
- 2019-08-09 00:47: update #40
- 2019-08-08 21:26: update #41
- 2019-08-10 00:34: update #42
- 2019-08-10 03:14: update #43
- 2019-08-09 18:15: update #44
- 2019-08-10 02:21: update #45
- 2019-08-13 16:54: update #46
- 2019-08-15 03:21: update #47
- 2019-08-17 03:17: update #48
- 2019-08-17 00:43: update #49
- 2019-08-18 03:46: update #50
- 2019-08-18 22:22: update #51
- 2019-08-19 19:49: update #52
- 2019-08-22 23:37: update #53
- 2019-08-25 06:25: update #54
- 2019-08-27 22:28: update #55
- 2019-08-27 17:36: update #56
- 2019-08-28 18:41: update #57
- 2019-08-30 01:00: update #58
- 2019-08-29 20:40: update #59
- 2019-08-30 22:56: update #60
- 2019-08-31 20:00: update #61
- 2019-09-02 06:56: update #62
- 2019-09-03 01:06: update #63
- 2019-09-04 03:59: update #64
- 2019-09-06 03:24: update #65
- 2019-09-07 01:59: update #66
- 2019-09-07 02:47: update #67
- 2019-09-07 07:11: update #68
- 2019-09-08 20:40: update #69
- 2019-09-10 04:29: update #70
- 2019-09-12 20:32: update #71
- 2019-09-15 18:18: update #72
- 2019-09-16 19:43: update #73
- 2019-09-16 21:14: update #74
- 2019-09-19 22:14: update #75
- 2019-09-20 23:58: update #76
- 2019-09-21 06:45: update #77
- 2019-09-23 22:46: update #78
- 2019-09-25 02:10: update #79
- 2019-09-24 19:03: update #80
- 2019-09-26 04:53: update #81
- 2019-09-27 16:21: update #82
- 2019-09-28 00:53: update #83
- 2019-09-29 03:18: update #84
- 2019-09-29 04:11: update #85
- 2019-09-30 03:58: update #86
- 2019-10-01 00:07: update #87
- 2019-10-01 03:34: update #88
- 2019-10-01 22:52: update #89
- 2019-10-03 22:31: update #90
- 2019-10-06 17:22: update #91
- 2019-10-08 16:20: update #92
- 2019-10-09 23:12: update #93
- 2019-10-10 20:02: update #94
- 2019-10-11 06:46: update #95
- 2019-10-11 05:53: update #96
- 2019-10-11 04:52: update #97
- 2019-10-11 21:39: update #98
- 2019-10-13 04:24: update #99
- 2019-10-18 04:01: update #100
- 2019-10-18 03:11: update #101
- 2019-10-17 17:48: update #102
- 2019-10-17 23:21: update #103
- 2019-10-18 20:22: update #104
- 2019-10-20 16:28: update #105
- 2019-10-23 18:00: update #106
- 2019-10-23 19:50: update #107
- 2019-10-23 19:30: update #108
- 2019-10-24 18:50: update #109
- 2019-10-25 16:45: update #110
- 2019-10-26 04:08: update #111
- 2019-10-26 01:18: update #112
- 2019-10-26 22:42: update #113
- 2019-10-27 20:34: update #114
- 2019-10-29 18:59: update #115
- 2019-10-29 22:28: update #116
- 2019-10-29 20:54: update #117
- 2019-10-30 01:44: update #118
- 2019-10-31 01:52: update #119
- 2019-10-30 16:38: update #120
- 2019-11-01 06:02: update #121
- 2019-11-01 06:52: update #122
- 2019-10-31 19:03: update #123
- 2019-11-01 02:56: update #124
- 2019-11-01 19:27: update #125
- 2019-11-02 00:39: update #126
- 2019-11-04 02:00: update #127
- 2019-11-04 21:25: update #128
- 2019-11-05 19:25: update #129
- 2019-11-07 02:19: update #130
- 2019-11-11 23:37: update #131
- 2019-11-13 01:58: update #132
- 2019-11-13 23:28: update #133
- 2019-11-15 05:11: update #134
- 2019-11-15 22:40: update #135
- 2019-11-17 00:19: update #136
- 2019-11-19 00:42: update #137
- 2019-11-20 05:11: update #138
- 2019-11-19 22:03: update #139
- 2019-11-19 18:42: update #140
- 2019-11-19 18:54: update #141
- 2019-11-20 17:49: update #142
- 2019-11-21 00:51: update #143
- 2019-11-21 20:56: update #144
- 2019-11-23 02:10: update #145
- 2019-11-22 22:55: update #146
- 2019-11-22 16:26: update #147
- 2019-11-22 20:50: update #148
- 2019-11-23 22:19: update #149
- 2019-11-26 02:30: update #150
- 2019-11-27 23:57: update #151
- 2019-11-29 00:15: update #152
- 2019-11-30 02:54: update #153
- 2019-11-29 16:37: update #154
- 2019-11-29 16:20: update #155
- 2019-11-30 23:02: update #156
- 2019-12-01 03:40: update #157
- 2019-12-04 03:37: update #158
- 2019-12-04 21:00: update #159
- 2019-12-06 18:59: update #160
- 2019-12-08 06:28: update #161
- 2019-12-10 17:05: update #162
- 2019-12-12 00:12: update #163
- 2019-12-13 05:54: update #164
- 2019-12-14 23:28: update #165
- 2019-12-17 01:54: update #166
- 2019-12-18 04:57: update #167
- 2019-12-19 00:31: update #168
- 2019-12-19 02:07: update #169
- 2019-12-20 05:32: update #170
- 2019-12-19 16:51: update #171
- 2019-12-19 22:31: update #172
- 2019-12-20 17:11: update #173
- 2019-12-23 17:40: update #174
- 2019-12-26 03:23: update #175
- 2019-12-25 19:06: update #176
- 2019-12-25 19:03: update #177
- 2019-12-26 23:41: update #178
- 2019-12-27 19:36: update #179
- 2019-12-30 04:36: update #180
- 2019-12-30 18:02: update #181
- 2019-12-31 02:30: update #182
- 2019-12-31 17:58: update #183
- 2019-12-31 21:56: update #184
- 2020-01-03 03:00: update #185
- 2020-01-04 05:29: update #186
- 2020-01-04 19:03: update #187
- 2020-01-06 17:41: update #188
- 2020-01-08 00:34: update #189
- 2020-01-08 01:52: update #190

tests/integration.rs
use std::process::Command;

fn netprobe_bin() -> Command {
    Command::new(env!("CARGO_BIN_EXE_netprobe"))
}

#[test]
fn test_help_output() {
    let output = netprobe_bin().arg("--help").output().expect("Failed to run netprobe");
    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.contains("netprobe"));
    assert!(stdout.contains("network diagnostics"));
}

#[test]
fn test_version_output() {
    let output = netprobe_bin()
        .arg("--version")
        .output()
        .expect("Failed to run netprobe");
    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.contains("netprobe"));
}

#[test]
fn test_ping_help() {
    let output = netprobe_bin()
        .args(&["ping", "--help"])
        .output()
        .expect("Failed to run netprobe");
    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.contains("ICMP echo requests"));
    assert!(stdout.contains("--count"));
    assert!(stdout.contains("--timeout"));
}

#[test]
fn test_scan_help() {
    let output = netprobe_bin()
        .args(&["scan", "--help"])
        .output()
        .expect("Failed to run netprobe");
    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.contains("TCP ports"));
    assert!(stdout.contains("--ports"));
    assert!(stdout.contains("--concurrency"));
}

#[test]
fn test_dns_help() {
    let output = netprobe_bin()
        .args(&["dns", "--help"])
        .output()
        .expect("Failed to run netprobe");
    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.contains("DNS"));
    assert!(stdout.contains("--record-type"));
}

#[test]
fn test_trace_help() {
    let output = netprobe_bin()
        .args(&["trace", "--help"])
        .output()
        .expect("Failed to run netprobe");
    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.contains("route"));
    assert!(stdout.contains("--max-hops"));
}

#[test]
fn test_http_help() {
    let output = netprobe_bin()
        .args(&["http", "--help"])
        .output()
        .expect("Failed to run netprobe");
    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.contains("HTTP"));
    assert!(stdout.contains("--requests"));
    assert!(stdout.contains("--concurrency"));
}

#[test]
fn test_no_subcommand_shows_help() {
    let output = netprobe_bin().output().expect("Failed to run netprobe");
    // Without a subcommand, clap should error
    assert!(!output.status.success());
    let stderr = String::from_utf8_lossy(&output.stderr);
    assert!(
        stderr.contains("SUBCOMMAND") || stderr.contains("subcommand") || stderr.contains("USAGE")
    );
}

#[test]
fn test_json_format_flag() {
    let output = netprobe_bin()
        .args(&["--format", "json", "ping", "--help"])
        .output()
        .expect("Failed to run netprobe");
    assert!(output.status.success());
}

#[test]
fn test_scan_localhost_closed_port() {
    // Scan a high port that should be closed
    let output = netprobe_bin()
        .args(&[
            "scan",
            "127.0.0.1",
            "--ports",
            "59123",
            "--timeout",
            "200",
        ])
        .output()
        .expect("Failed to run netprobe");
    // Should complete without panic
    assert!(output.status.success());
}

#[test]
fn test_scan_invalid_port_range() {
    let output = netprobe_bin()
        .args(&["scan", "127.0.0.1", "--ports", "abc"])
        .output()
        .expect("Failed to run netprobe");
    assert!(!output.status.success());
}


src/dns.rs
use crate::lib::NetprobeError;
use crate::output::{self, OutputFormat};
use anyhow::Result;
use colored::*;
use serde::Serialize;
use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use std::str::FromStr;
use std::time::{Duration, Instant};
use trust_dns_resolver::config::{
    NameServerConfig, Protocol, ResolverConfig, ResolverOpts,
};
use trust_dns_resolver::proto::rr::RecordType;
use trust_dns_resolver::TokioAsyncResolver;

#[derive(Debug, Serialize)]
struct DnsResult {
    domain: String,
    record_type: String,
    server: String,
    query_time_ms: f64,
    records: Vec<DnsRecord>,
}

#[derive(Debug, Serialize)]
struct DnsRecord {
    name: String,
    record_type: String,
    ttl: u32,
    value: String,
}

pub async fn run(
    domain: &str,
    record_type_str: &str,
    server: Option<&str>,
    timeout: Duration,
    format: OutputFormat,
) -> Result<()> {
    let record_type = parse_record_type(record_type_str)?;

    // Build resolver configuration
    let (resolver_config, server_name) = if let Some(srv) = server {
        let ip = IpAddr::from_str(srv)
            .map_err(|_| NetprobeError::ParseError(format!("Invalid DNS server: {}", srv)))?;
        let ns = NameServerConfig {
            socket_addr: SocketAddr::new(ip, 53),
            protocol: Protocol::Udp,
            tls_dns_name: None,
            trust_nx_responses: false,
            bind_addr: None,
        };
        let mut config = ResolverConfig::new();
        config.add_name_server(ns);
        (config, srv.to_string())
    } else {
        (
            ResolverConfig::default(),
            "system default".to_string(),
        )
    };

    let mut opts = ResolverOpts::default();
    opts.timeout = timeout;
    opts.attempts = 3;
    opts.use_hosts_file = false;

    let resolver = TokioAsyncResolver::tokio(resolver_config, opts)?;

    if format == OutputFormat::Text {
        println!(
            "{} Querying {} record for {} via {}",
            "DNS".green().bold(),
            record_type_str.to_uppercase().yellow(),
            domain.bold(),
            server_name.cyan()
        );
        println!();
    }

    let start = Instant::now();
    let records = perform_lookup(&resolver, domain, record_type, record_type_str).await?;
    let query_time = start.elapsed();

    if format == OutputFormat::Text {
        if records.is_empty() {
            println!("{}", "No records found.".yellow());
        } else {
            println!(
                "{:<40} {:<8} {:<8} {}",
                "NAME".bold(),
                "TYPE".bold(),
                "TTL".bold(),
                "VALUE".bold()
            );
            println!("{}", "-".repeat(80));

            for record in &records {
                println!(
                    "{:<40} {:<8} {:<8} {}",
                    record.name.white(),
                    record.record_type.cyan(),
                    record.ttl.to_string().dimmed(),
                    record.value.green()
                );
            }
        }

        println!();
        println!(
            "Query time: {}",
            format!("{:.2}ms", query_time.as_secs_f64() * 1000.0).yellow()
        );
        println!("Server: {}", server_name.cyan());
    } else {
        let result = DnsResult {
            domain: domain.to_string(),
            record_type: record_type_str.to_uppercase(),
            server: server_name,
            query_time_ms: query_time.as_secs_f64() * 1000.0,
            records,
        };
        output::print_json(&result)?;
    }

    Ok(())
}

fn parse_record_type(s: &str) -> Result<RecordType, NetprobeError> {
    match s.to_uppercase().as_str() {
        "A" => Ok(RecordType::A),
        "AAAA" => Ok(RecordType::AAAA),
        "MX" => Ok(RecordType::MX),
        "CNAME" => Ok(RecordType::CNAME),
        "TXT" => Ok(RecordType::TXT),
        "NS" => Ok(RecordType::NS),
        "SOA" => Ok(RecordType::SOA),
        "PTR" => Ok(RecordType::PTR),
        "SRV" => Ok(RecordType::SRV),
        _ => Err(NetprobeError::InvalidRecordType(s.to_string())),
    }
}

async fn perform_lookup(
    resolver: &TokioAsyncResolver,
    domain: &str,
    record_type: RecordType,
    type_str: &str,
) -> Result<Vec<DnsRecord>> {
    let mut records = Vec::new();
    let name = trust_dns_resolver::Name::from_str(domain)
        .map_err(|e| NetprobeError::ParseError(format!("Invalid domain name: {}", e)))?;

    match record_type {
        RecordType::A => {
            let lookup = resolver.ipv4_lookup(domain).await?;
            for addr in lookup.iter() {
                records.push(DnsRecord {
                    name: domain.to_string(),
                    record_type: "A".to_string(),
                    ttl: lookup.as_lookup().record_iter().next().map(|r| r.ttl()).unwrap_or(0),
                    value: addr.to_string(),
                });
            }
        }
        RecordType::AAAA => {
            let lookup = resolver.ipv6_lookup(domain).await?;
            for addr in lookup.iter() {
                records.push(DnsRecord {
                    name: domain.to_string(),
                    record_type: "AAAA".to_string(),
                    ttl: lookup.as_lookup().record_iter().next().map(|r| r.ttl()).unwrap_or(0),
                    value: addr.to_string(),
                });
            }
        }
        RecordType::MX => {
            let lookup = resolver.mx_lookup(domain).await?;
            for mx in lookup.iter() {
                records.push(DnsRecord {
                    name: domain.to_string(),
                    record_type: "MX".to_string(),
                    ttl: lookup.as_lookup().record_iter().next().map(|r| r.ttl()).unwrap_or(0),
                    value: format!("{} {}", mx.preference(), mx.exchange()),
                });
            }
        }
        RecordType::TXT => {
            let lookup = resolver.txt_lookup(domain).await?;
            for txt in lookup.iter() {
                let txt_data: Vec<String> = txt.iter().map(|d| {
                    String::from_utf8_lossy(d).to_string()
                }).collect();
                records.push(DnsRecord {
                    name: domain.to_string(),
                    record_type: "TXT".to_string(),
                    ttl: lookup.as_lookup().record_iter().next().map(|r| r.ttl()).unwrap_or(0),
                    value: txt_data.join(""),
                });
            }
        }
        RecordType::NS => {
            let lookup = resolver.ns_lookup(domain).await?;
            for ns in lookup.iter() {
                records.push(DnsRecord {
                    name: domain.to_string(),
                    record_type: "NS".to_string(),
                    ttl: lookup.as_lookup().record_iter().next().map(|r| r.ttl()).unwrap_or(0),
                    value: ns.to_string(),
                });
            }
        }
        RecordType::SOA => {
            let lookup = resolver.soa_lookup(domain).await?;
            for soa in lookup.iter() {
                records.push(DnsRecord {
                    name: domain.to_string(),
                    record_type: "SOA".to_string(),
                    ttl: lookup.as_lookup().record_iter().next().map(|r| r.ttl()).unwrap_or(0),
                    value: format!(
                        "{} {} {} {} {} {} {}",
                        soa.mname(),
                        soa.rname(),
                        soa.serial(),
                        soa.refresh(),
                        soa.retry(),
                        soa.expire(),
                        soa.minimum()
                    ),
                });
            }
        }
        _ => {
            let lookup = resolver.lookup(name, record_type).await?;
            for record in lookup.record_iter() {
                records.push(DnsRecord {
                    name: domain.to_string(),
                    record_type: type_str.to_uppercase(),
                    ttl: record.ttl(),
                    value: format!("{:?}", record.data()),
                });
            }
        }
    }

    Ok(records)
}
// fix aaaa edge case


src/lib.rs
use serde::Serialize;
use std::net::{IpAddr, ToSocketAddrs};
use std::time::Duration;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum NetprobeError {
    #[error("DNS resolution failed for host '{0}'")]
    DnsResolutionFailed(String),

    #[error("Connection timed out after {0:?}")]
    Timeout(Duration),

    #[error("Socket error: {0}")]
    SocketError(String),

    #[error("Invalid port range: {0}")]
    InvalidPortRange(String),

    #[error("Invalid record type: {0}")]
    InvalidRecordType(String),

    #[error("Permission denied: {0}")]
    PermissionDenied(String),

    #[error("HTTP error: {0}")]
    HttpError(String),

    #[error("Parse error: {0}")]
    ParseError(String),
}

/// Resolve a hostname to its first IP address.
pub fn resolve_host(host: &str) -> Result<IpAddr, NetprobeError> {
    // Try parsing as an IP address first
    if let Ok(ip) = host.parse::<IpAddr>() {
        return Ok(ip);
    }

    // Resolve hostname
    let addr = format!("{}:0", host);
    addr.to_socket_addrs()
        .map_err(|_| NetprobeError::DnsResolutionFailed(host.to_string()))?
        .next()
        .map(|socket_addr| socket_addr.ip())
        .ok_or_else(|| NetprobeError::DnsResolutionFailed(host.to_string()))
}

/// Parse a port range specification into a vector of ports.
///
/// Supported formats:
/// - Single port: "80"
/// - Range: "1-1024"
/// - Comma-separated: "80,443,8080"
/// - Mixed: "22,80,443,8000-8100"
pub fn parse_ports(spec: &str) -> Result<Vec<u16>, NetprobeError> {
    let mut ports = Vec::new();

    for part in spec.split(',') {
        let part = part.trim();
        if part.contains('-') {
            let bounds: Vec<&str> = part.split('-').collect();
            if bounds.len() != 2 {
                return Err(NetprobeError::InvalidPortRange(part.to_string()));
            }
            let start: u16 = bounds[0]
                .trim()
                .parse()
                .map_err(|_| NetprobeError::InvalidPortRange(part.to_string()))?;
            let end: u16 = bounds[1]
                .trim()
                .parse()
                .map_err(|_| NetprobeError::InvalidPortRange(part.to_string()))?;
            if start > end {
                return Err(NetprobeError::InvalidPortRange(format!(
                    "start ({}) > end ({})",
                    start, end
                )));
            }
            for port in start..=end {
                ports.push(port);
            }
        } else {
            let port: u16 = part
                .parse()
                .map_err(|_| NetprobeError::InvalidPortRange(part.to_string()))?;
            ports.push(port);
        }
    }

    if ports.is_empty() {
        return Err(NetprobeError::InvalidPortRange(spec.to_string()));
    }

    Ok(ports)
}

/// Statistics computed from a sequence of latency samples.
#[derive(Debug, Clone, Serialize)]
pub struct LatencyStats {
    pub min_ms: f64,
    pub max_ms: f64,
    pub avg_ms: f64,
    pub stddev_ms: f64,
    pub jitter_ms: f64,
    pub median_ms: f64,
    pub p95_ms: f64,
    pub p99_ms: f64,
    pub samples: usize,
}

impl LatencyStats {
    /// Compute statistics from a vector of latency durations.
    pub fn from_durations(durations: &[Duration]) -> Option<Self> {
        if durations.is_empty() {
            return None;
        }

        let mut ms_values: Vec<f64> = durations.iter().map(|d| d.as_secs_f64() * 1000.0).collect();
        ms_values.sort_by(|a, b| a.partial_cmp(b).unwrap());

        let n = ms_values.len();
        let min_ms = ms_values[0];
        let max_ms = ms_values[n - 1];
        let sum: f64 = ms_values.iter().sum();
        let avg_ms = sum / n as f64;

        let variance = ms_values.iter().map(|v| (v - avg_ms).powi(2)).sum::<f64>() / n as f64;
        let stddev_ms = variance.sqrt();

        // Jitter = mean of absolute differences between consecutive samples
        let jitter_ms = if n > 1 {
            let jitter_sum: f64 = ms_values
                .windows(2)
                .map(|w| (w[1] - w[0]).abs())
                .sum();
            jitter_sum / (n - 1) as f64
        } else {
            0.0
        };

        let median_ms = if n % 2 == 0 {
            (ms_values[n / 2 - 1] + ms_values[n / 2]) / 2.0
        } else {
            ms_values[n / 2]
        };

        let p95_ms = percentile(&ms_values, 95.0);
        let p99_ms = percentile(&ms_values, 99.0);

        Some(LatencyStats {
            min_ms,
            max_ms,
            avg_ms,
            stddev_ms,
            jitter_ms,
            median_ms,
            p95_ms,
            p99_ms,
            samples: n,
        })
    }
}

/// Compute the k-th percentile from a sorted slice.
fn percentile(sorted: &[f64], pct: f64) -> f64 {
    if sorted.is_empty() {
        return 0.0;
    }
    if sorted.len() == 1 {
        return sorted[0];
    }
    let rank = (pct / 100.0) * (sorted.len() - 1) as f64;
    let lower = rank.floor() as usize;
    let upper = rank.ceil() as usize;
    let frac = rank - lower as f64;
    sorted[lower] * (1.0 - frac) + sorted[upper] * frac
}

/// Format a duration as a human-readable string.
pub fn format_duration(d: Duration) -> String {
    let ms = d.as_secs_f64() * 1000.0;
    if ms < 1.0 {
        format!("{:.1}us", ms * 1000.0)
    } else if ms < 1000.0 {
        format!("{:.2}ms", ms)
    } else {
        format!("{:.2}s", ms / 1000.0)
    }
}

/// Well-known port to service name mapping for common services.
pub fn service_name(port: u16) -> Option<&'static str> {
    match port {
        20 => Some("ftp-data"),
        21 => Some("ftp"),
        22 => Some("ssh"),
        23 => Some("telnet"),
        25 => Some("smtp"),
        53 => Some("dns"),
        67 => Some("dhcp-server"),
        68 => Some("dhcp-client"),
        80 => Some("http"),
        110 => Some("pop3"),
        119 => Some("nntp"),
        123 => Some("ntp"),
        135 => Some("msrpc"),
        139 => Some("netbios"),
        143 => Some("imap"),
        161 => Some("snmp"),
        194 => Some("irc"),
        389 => Some("ldap"),
        443 => Some("https"),
        445 => Some("smb"),
        465 => Some("smtps"),
        514 => Some("syslog"),
        587 => Some("submission"),
        636 => Some("ldaps"),
        993 => Some("imaps"),
        995 => Some("pop3s"),
        1433 => Some("mssql"),
        1521 => Some("oracle"),
        3306 => Some("mysql"),
        3389 => Some("rdp"),
        5432 => Some("postgresql"),
        5672 => Some("amqp"),
        5900 => Some("vnc"),
        6379 => Some("redis"),
        8080 => Some("http-alt"),
        8443 => Some("https-alt"),
        9200 => Some("elasticsearch"),
        9300 => Some("elasticsearch-transport"),
        11211 => Some("memcached"),
        27017 => Some("mongodb"),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_single_port() {
        let ports = parse_ports("80").unwrap();
        assert_eq!(ports, vec![80]);
    }

    #[test]
    fn test_parse_port_range() {
        let ports = parse_ports("1-5").unwrap();
        assert_eq!(ports, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_parse_port_list() {
        let ports = parse_ports("22,80,443").unwrap();
        assert_eq!(ports, vec![22, 80, 443]);
    }

    #[test]
    fn test_parse_mixed_ports() {
        let ports = parse_ports("22,80-82,443").unwrap();
        assert_eq!(ports, vec![22, 80, 81, 82, 443]);
    }

    #[test]
    fn test_parse_invalid_port() {
        assert!(parse_ports("abc").is_err());
    }

    #[test]
    fn test_parse_reversed_range() {
        assert!(parse_ports("100-50").is_err());
    }

    #[test]
    fn test_latency_stats() {
        let durations: Vec<Duration> = vec![10, 20, 30, 40, 50]
            .into_iter()
            .map(Duration::from_millis)
            .collect();
        let stats = LatencyStats::from_durations(&durations).unwrap();
        assert!((stats.min_ms - 10.0).abs() < 0.1);
        assert!((stats.max_ms - 50.0).abs() < 0.1);
        assert!((stats.avg_ms - 30.0).abs() < 0.1);
        assert!((stats.median_ms - 30.0).abs() < 0.1);
    }

    #[test]
    fn test_latency_stats_empty() {
        assert!(LatencyStats::from_durations(&[]).is_none());
    }

    #[test]
    fn test_format_duration() {
        assert!(format_duration(Duration::from_micros(500)).contains("us"));
        assert!(format_duration(Duration::from_millis(50)).contains("ms"));
        assert!(format_duration(Duration::from_secs(2)).contains("s"));
    }

    #[test]
    fn test_service_name() {
        assert_eq!(service_name(80), Some("http"));
        assert_eq!(service_name(443), Some("https"));
        assert_eq!(service_name(12345), None);
    }

    #[test]
    fn test_resolve_ip_address() {
        let ip = resolve_host("127.0.0.1").unwrap();
        assert_eq!(ip, "127.0.0.1".parse::<IpAddr>().unwrap());
    }
}
// refactored error types
